### ISSUE TITLE
Resource `git` actions `checkout` and `sync` abend in a Dokken environment: no /dev/tty

### ISSUE THREAD

**Issue by taqtiqa-mark** [chef/chef#8855]
## Description

Resource `git` actions `checkout` and `sync` abend in a Dokken environment: no /dev/tty
 
## Chef Version

```
Chef Development Kit Version: 3.11.3
chef-client version: 14.13.11
delivery version: master (9d07501a3b347cc687c902319d23dc32dd5fa621)
berks version: 7.0.8
kitchen version: 1.25.0
inspec version: 3.9.3
```
## Platform Version

Ubuntu 18.04

## Replication Case

When Dokken testing the git resource as follows:

```
33:   git "Install #{new_resource.name} plugin" do
34:     destination ::File.join(root_path, 'plugins', new_resource.name)
35:     repository new_resource.git_url
36:     reference new_resource.git_ref
37:     user new_resource.user if new_resource.user
38:     action :checkout
39:   end
40: end
```

## Client Output

I encounter an issue as described on [SO](https://stackoverflow.com/a/57458443/152860).

The error is:

```
* git[Install julia-build plugin] action checkout

================================================================================
Error executing action `checkout` on resource 'git[Install julia-build plugin]'
================================================================================

Mixlib::ShellOut::ShellCommandFailed
------------------------------------
Expected process to exit with [0], but received '128'
---- Begin output of git ls-remote "https://github.com/jlenv/jlenv-build.git" "master*" ----
STDOUT: 
STDERR: fatal: could not read Username for 'https://github.com': No such device or address
---- End output of git ls-remote "https://github.com/jlenv/jlenv-build.git" "master*" ----
Ran git ls-remote "https://github.com/jlenv/jlenv-build.git" "master*" returned 128

```
I was able to test and confirm that in the Dokken environment:
```
echo ahoj > /dev/tty
bash: /dev/tty: No such device or address
```

It appears the issue is [here](https://github.com/chef/chef/blob/7558c414030d082194d0d2b5f74279ce151517e9/lib/chef/provider/git.rb#L298)?

## Stacktrace

As Above.

## Workaround

A potential workaround is to use the [git-cookbook](https://github.com/chef-cookbooks/git), which does not use `git ls-remote ...`.  
Will update if that succeeds.

**Comment by taqtiqa-mark** [comment]
I believe the correct way to accomplish this is `git show-ref --quiet --verify <ref> >/dev/null 2>&1`, redirecting stderr and stdout to `/dev/null` and returning the result depending on the exit code?

Before I sink more of my time. Would you accept a pull request implementing this approach?

**Comment by lamont-granquist** [comment]
yeah although `2>&1` is a bash-ism and it needs to work if the shell is tcsh, and probably needs to work on windows as well.

if there's a way to pass the moral equivalent of ssh's `-n` argument to git that would be better rather than trying to convince mixlib-shellout to wire up /dev/null in a shel-portable-fashion and do <something that i couldn't guess> on windows, which is probably going to be a lot harder.

**Comment by taqtiqa-mark** [comment]
@lamont-granquist thanks for looking at this.
Agreed.  I've been thinking about this a little more.  

The absence of a TTY is the default in many of my chef-automated use cases i.e. cloud server setup's, containers, etc.  
From my perspective (and reading around), issues such as this one are circumvented by using vagrant, passing docker a `--device` or avoiding such stdout/stderror parsing recipes.  
In my case I'm making more use of dokken, and here I can't sidestep the git resource.

Rather than piping to stdout and stderr it seems the more generic approach is to send the data to files and read it from there. 
Agreed?

Is there an example of the best-practice Chef way to write/read temp files in a recipe?

In this case the process would be `git show-ref --quiet --verify` to get the result/return code.  
If the return indicates verified then read in the output file generated.

This (no tty) seems to be an issue more widespread than the `git` resource?
